context/cancel.go

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package context
+
+import "sync/atomic"
+
+type cancellable struct {
+	cancelled *int32
```

Comment:
In general yes, however there is an optimization made in the go compiler that says if the receiver type fits in a machine word, then it will stored directly in the `data` field: https://research.swtch.com/interfaces

If we use *int32 as the field and use `cancellable` as the receiver type, the size of the struct value is 8 bytes, which fits in a single machine word and therefore directly stored in the `data` field. The value contains an *int32 that points to a 4-byte memory chunk on the heap.

If we use int32 as the field and use `*cancellable` as the receiver type, the size of the struct pointer is 8 bytes, which also fits in a single machine word and stored in the `data` field. The struct pointer then points to a struct allocated on the heap containing an `int32` object.

Therefore, the two approaches are essentially equivalent.